src/ast2ram/seminaive/ClauseTranslator.cpp

```diff
@@ -269,38 +269,66 @@ Own<ram::Operation> ClauseTranslator::addAdtUnpack(
         Own<ram::Operation> op, const ast::BranchInit* adt, int curLevel) const {
     assert(!context.isADTEnum(adt) && "ADT enums should not be unpacked");
 
-    // set branch tag constraint
-    op = addEqualityCheck(std::move(op), mk<ram::TupleElement>(curLevel, 0),
-            mk<ram::SignedConstant>(context.getADTBranchId(adt)), false);
+    auto args = adt->getArguments();
+
+    // direct encoding for a single/no argument
+    if (args.size() <= 1) {
+        // add remaining constant constraints
+        auto dummyArg = mk<ast::UnnamedVariable>();
+        std::vector<ast::Argument*> branchArguments;
+        branchArguments.push_back(dummyArg.get());
+        for (auto* arg : args) {
+            branchArguments.push_back(arg);
+        }
+        op = addConstantConstraints(curLevel, branchArguments, std::move(op));
+
+        // set branch tag constraint
+        op = addEqualityCheck(std::move(op), mk<ram::TupleElement>(curLevel, 0),
+                mk<ram::SignedConstant>(context.getADTBranchId(adt)), false);
+
+        // add an unpack level
+        const Location& loc = valueIndex->getDefinitionPoint(*adt);
+        op = mk<ram::UnpackRecord>(std::move(op), curLevel, makeRamTupleElement(loc), args.size() + 1);
+
+    } else {
+        // add remaining constant constraints
+        std::vector<ast::Argument*> branchArguments;
+        for (auto* arg : args) {
+            branchArguments.push_back(arg);
+        }
+        op = addConstantConstraints(curLevel + 1, branchArguments, std::move(op));
+
+        // Unpack payload
+        op = mk<ram::UnpackRecord>(
+                std::move(op), curLevel + 1, mk<ram::TupleElement>(curLevel, 1), args.size());
 
-    // add remaining constant constraints
-    auto dummyArg = mk<ast::UnnamedVariable>();
-    std::vector<ast::Argument*> branchArguments;
-    branchArguments.push_back(dummyArg.get());
-    for (auto* arg : adt->getArguments()) {
-        branchArguments.push_back(arg);
+        // set branch tag constraint
+        op = addEqualityCheck(std::move(op), mk<ram::TupleElement>(curLevel, 0),
+                mk<ram::SignedConstant>(context.getADTBranchId(adt)), false);
+
+        // add an unpack level
+        const Location& loc = valueIndex->getDefinitionPoint(*adt);
+        op = mk<ram::UnpackRecord>(std::move(op), curLevel, makeRamTupleElement(loc), 2);
     }
-    op = addConstantConstraints(curLevel, branchArguments, std::move(op));
 
-    // add an unpack level
-    const Location& loc = valueIndex->getDefinitionPoint(*adt);
-    op = mk<ram::UnpackRecord>(std::move(op), curLevel, makeRamTupleElement(loc), branchArguments.size());
     return op;
 }
 
 Own<ram::Operation> ClauseTranslator::addVariableIntroductions(
         const ast::Clause& clause, Own<ram::Operation> op) {
     for (int i = operators.size() - 1; i >= 0; i--) {
+        // ADT need two levels
```

Comment:
Instead of hacking it like this, can possibly just add an extra duplicate level during node indexing for ADTs - `indexNodeArguments` - and then double increment here.

Reply:
This patch (for the upstream branch) is a cleaner way the levels can be fixed:

```
diff --git a/src/ast2ram/seminaive/ClauseTranslator.cpp b/src/ast2ram/seminaive/ClauseTranslator.cpp
index 527ce632b..d8c1aefba 100644
--- a/src/ast2ram/seminaive/ClauseTranslator.cpp
+++ b/src/ast2ram/seminaive/ClauseTranslator.cpp
@@ -282,9 +282,15 @@ Own<ram::Operation> ClauseTranslator::addAdtUnpack(
     }
     op = addConstantConstraints(curLevel, branchArguments, std::move(op));
 
-    // add an unpack level
     const Location& loc = valueIndex->getDefinitionPoint(*adt);
-    op = mk<ram::UnpackRecord>(std::move(op), curLevel, makeRamTupleElement(loc), branchArguments.size());
+
+    // add an unpack level for arguments
+    op = mk<ram::UnpackRecord>(std::move(op), curLevel + 1, mk<ram::TupleElement>(curLevel, 1), branchArguments.size());
+
+    // add an unpack level for main record
+    op = mk<ram::UnpackRecord>(std::move(op), curLevel, makeRamTupleElement(loc), 2);
+
+
     return op;
 }
 
@@ -301,6 +307,7 @@ Own<ram::Operation> ClauseTranslator::addVariableIntroductions(
         } else if (const auto* adt = as<ast::BranchInit>(curOp)) {
             // add adt arguments through an unpack
             op = addAdtUnpack(std::move(op), adt, i);
+            i--;
         } else {
             fatal("Unsupported AST node for creation of scan-level!");
         }
@@ -645,15 +652,10 @@ void ClauseTranslator::indexNodeArguments(int nodeLevel, const std::vector<ast::
             valueIndex->setAdtDefinition(*adt, nodeLevel, i);
 
             // introduce new nesting level for unpack
-            auto unpackLevel = addOperatorLevel(adt);
+            /* auto mainUnpackLevel = */ addOperatorLevel(adt);
+            auto argumentUnpackLevel = addOperatorLevel(adt);
 
-            auto dummyArg = mk<ast::UnnamedVariable>();
-            std::vector<ast::Argument*> arguments;
-            arguments.push_back(dummyArg.get());
-            for (auto* arg : adt->getArguments()) {
-                arguments.push_back(arg);
-            }
-            indexNodeArguments(unpackLevel, arguments);
+            indexNodeArguments(argumentUnpackLevel, adt->getArguments());
         }
     }
 }
```

This is just the core idea - would still need variable references to be fixed when referring to arguments, etc.